File: kotlin/src/main/kotlin/com/thinkoff/userintent/IntentClient.kt
// SPDX-License-Identifier: AGPL-3.0

package com.thinkoff.userintent

import java.net.HttpURLConnection
import java.net.URL
import kotlinx.coroutines.*

/**
 * REST client for the User Intent API.
 * For Wear OS, Android, and JVM-based agents.
 */
class IntentClient(
    baseUrl: String,
    private val apiKey: String,
    private val userId: String,
    private val deviceId: String? = null,
    private val timeoutMs: Int = 10000,
    private val heartbeatIntervalMs: Long = 30000L
) {
    private val baseUrl = baseUrl.trimEnd('/')
    private var heartbeatJob: Job? = null

    // --- Profile ---

    suspend fun getProfile(): Map<String, Any?> =
        request("GET", "/profile/$userId")

    // --- Intent ---

    suspend fun getIntent(): Map<String, Any?> =
        request("GET", "/intent/$userId")

    suspend fun patchDevice(fields: Map<String, Any?>): Map<String, Any?> {
        val id = deviceId ?: throw IllegalStateException("No deviceId configured")
        return request("PATCH", "/intent/$userId/$id", fields)
    }

    suspend fun patchAgent(name: String, fields: Map<String, Any?>): Map<String, Any?> =
        request("PATCH", "/intent/$userId/agents/$name", fields)

    suspend fun removeDevice(id: String? = null): Map<String, Any?> {
        val target = id ?: deviceId ?: throw IllegalStateException("No deviceId configured")
        return request("DELETE", "/intent/$userId/$target")
    }

    // --- Derived state helpers ---

    @Suppress("UNCHECKED_CAST")
    suspend fun getDerived(): Map<String, Any?> {
        val intent = getIntent()
        return intent["derived"] as? Map<String, Any?> ?: emptyMap()
    }

    suspend fun isInMeeting(): Boolean =
        getDerived()["urgency_mode"] == "text-only"

    suspend fun shouldSuppressAudio(): Boolean =
        getDerived()["suppress_audio"] == true

    suspend fun preferredDevice(): String? =
        getDerived()["preferred_device"] as? String

    // --- Heartbeat ---

    suspend fun heartbeat() {
        val id = deviceId ?: throw IllegalStateException("No deviceId configured")
        request("PATCH", "/intent/$userId/$id", mapOf("heartbeat" to true))
    }

    fun startHeartbeat(scope: CoroutineScope) {
        stopHeartbeat()
        heartbeatJob = scope.launch {
            while (isActive) {
                try { heartbeat() } catch (_: Exception) {}
                delay(maxOf(heartbeatIntervalMs, 10000L))
            }
        }
    }

    fun stopHeartbeat() {
        heartbeatJob?.cancel()
        heartbeatJob = null
    }

    // --- Internal ---

    private suspend fun request(
        method: String,
        path: String,
        body: Map<String, Any?>? = null
    ): Map<String, Any?> = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl$path")
        val conn = url.openConnection() as HttpURLConnection
        conn.requestMethod = if (method == "PATCH") "POST" else method
        if (method == "PATCH") conn.setRequestProperty("X-HTTP-Method-Override", "PATCH")
        conn.connectTimeout = timeoutMs
        conn.readTimeout = timeoutMs
        conn.setRequestProperty("X-API-Key", apiKey)
        conn.setRequestProperty("Content-Type", "application/json")

        if (body != null && method != "GET" && method != "DELETE") {
            conn.doOutput = true
            conn.outputStream.use { os ->
                os.write(toJson(body).toByteArray())
            }
        }

        val code = conn.responseCode
        if (code !in 200..299) {
            val errorBody = try { conn.errorStream?.bufferedReader()?.readText() ?: "" } catch (_: Exception) { "" }
            throw IntentApiException(code, errorBody)
        }

        val responseText = try { conn.inputStream.bufferedReader().readText() } catch (_: Exception) { "" }
        if (responseText.isBlank()) return@withContext emptyMap()
        parseJson(responseText)
    }

    // Minimal JSON helpers (no external dependency)
    private fun toJson(map: Map<String, Any?>): String {
        val entries = map.entries.joinToString(",") { (k, v) ->
            "\"$k\":${valueToJson(v)}"
        }
        return "{$entries}"
    }

    private fun valueToJson(v: Any?): String = when (v) {
        null -> "null"
        is String -> "\"$v\""
        is Boolean, is Number -> "$v"
        is Map<*, *> -> toJson(@Suppress("UNCHECKED_CAST") (v as Map<String, Any?>))
        else -> "\"$v\""
    }

    @Suppress("UNCHECKED_CAST")
    private fun parseJson(text: String): Map<String, Any?> {
        // Delegate to org.json which is available on Android
        return try {
            val obj = org.json.JSONObject(text)
            obj.keys().asSequence().associate { it to obj.opt(it) }
        } catch (_: Exception) {
            emptyMap()
        }
    }
}

class IntentApiException(val statusCode: Int, val body: String) :
    Exception("Intent API error $statusCode: $body")

File: swift/Sources/UserIntentKit/IntentClient.swift
// SPDX-License-Identifier: AGPL-3.0

import Foundation

/// REST client for the User Intent API.
/// Talks to an Ant Farm instance hosting profile + intent endpoints.
public class IntentClient {
    private let baseUrl: String
    private let apiKey: String
    private let userId: String
    private let deviceId: String?
    private let timeoutInterval: TimeInterval
    private var heartbeatTimer: Timer?

    public init(
        baseUrl: String,
        apiKey: String,
        userId: String,
        deviceId: String? = nil,
        timeoutInterval: TimeInterval = 10.0
    ) {
        self.baseUrl = baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        self.apiKey = apiKey
        self.userId = userId
        self.deviceId = deviceId
        self.timeoutInterval = timeoutInterval
    }

    // MARK: - Profile

    public func getProfile() async throws -> [String: Any] {
        return try await request(method: "GET", path: "/profile/\(userId)")
    }

    // MARK: - Intent

    public func getIntent() async throws -> [String: Any] {
        return try await request(method: "GET", path: "/intent/\(userId)")
    }

    public func patchDevice(_ fields: [String: Any]) async throws {
        guard let deviceId else { throw IntentError.noDeviceId }
        let _ = try await request(method: "PATCH", path: "/intent/\(userId)/\(deviceId)", body: fields)
    }

    public func patchAgent(name: String, fields: [String: Any]) async throws {
        let _ = try await request(method: "PATCH", path: "/intent/\(userId)/agents/\(name)", body: fields)
    }

    public func removeDevice(_ id: String? = nil) async throws {
        let target = id ?? deviceId
        guard let target else { throw IntentError.noDeviceId }
        let _ = try await request(method: "DELETE", path: "/intent/\(userId)/\(target)")
    }

    // MARK: - Derived state helpers

    public func getDerived() async throws -> [String: Any] {
        let intent = try await getIntent()
        return intent["derived"] as? [String: Any] ?? [:]
    }

    public func isInMeeting() async throws -> Bool {
        let derived = try await getDerived()
        return (derived["urgency_mode"] as? String) == "text-only"
    }

    public func shouldSuppressAudio() async throws -> Bool {
        let derived = try await getDerived()
        return (derived["suppress_audio"] as? Bool) ?? false
    }

    public func preferredDevice() async throws -> String? {
        let derived = try await getDerived()
        return derived["preferred_device"] as? String
    }

    // MARK: - Heartbeat

    public func heartbeat() async throws {
        guard let deviceId else { throw IntentError.noDeviceId }
        let _ = try await request(method: "PATCH", path: "/intent/\(userId)/\(deviceId)", body: ["heartbeat": true])
    }

    public func startHeartbeat(interval: TimeInterval = 30.0) {
        stopHeartbeat()
        heartbeatTimer = Timer.scheduledTimer(withTimeInterval: max(interval, 10.0), repeats: true) { [weak self] _ in
            Task {
                try? await self?.heartbeat()
            }
        }
    }

    public func stopHeartbeat() {
        heartbeatTimer?.invalidate()
        heartbeatTimer = nil
    }

    // MARK: - Internal

    private func request(method: String, path: String, body: [String: Any]? = nil) async throws -> [String: Any] {
        guard let url = URL(string: "\(baseUrl)\(path)") else {
            throw IntentError.invalidUrl
        }

        var req = URLRequest(url: url)
        req.httpMethod = method
        req.timeoutInterval = timeoutInterval
        req.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let body, method != "GET", method != "DELETE" {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        let (data, response) = try await URLSession.shared.data(for: req)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw IntentError.invalidResponse
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            let bodyText = String(data: data, encoding: .utf8) ?? ""
            throw IntentError.httpError(statusCode: httpResponse.statusCode, body: bodyText)
        }

        if data.isEmpty { return [:] }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return [:]
        }

        return json
    }
}

public enum IntentError: Error {
    case noDeviceId
    case invalidUrl
    case invalidResponse
    case httpError(statusCode: Int, body: String)
}
